File: tp2/ej6/ej.hs
type Nombre = String 

newtype DNI = DNI Int

{-
 - Haskell en cuanto a type, tiene equivalencia de tipos estructural, ya que 2 tipos 
 - son equivalentes cuando tienen la misma estructura. Esto lo podemos notar, ya que 
 - g toma como parametro un nombre, y sin embargo le pude pasar un String 
 -}

f :: String -> String 
f xs = g xs 

g :: Nombre -> Nombre 
g xs = xs

{-
 - Haskell en cuanto a newtype, tiene equivalencia de tipos por nombre, ya que 
 - 2 tipos son equivalentes cuando tienen el mismo nombre. Esto lo podemos notar,
 - ya que no puedo invocar g' con un Int 
 -}

f' :: Int -> Int 
f' x = g' x

g' :: DNI -> DNI 
g' (DNI n) = (DNI n)

{-
 - Entonces, Haskell maneja los dos enfoques de equivalencia de tipos, dependiendo 
 - la construccion que se use (type o newtype/data)
 -}

File: tp2/ej5/ej.c
#include <stdio.h>

struct Punto{
  int x;
  int y;
};

struct Coordenada{
  int x;
  int y;
};

struct Punto p1,p2;
struct Coordenada c;

int main(){
  p1.x = 1;
  p1.y = 2;
  p2 = p1;
  c = p1;
  /**
   * ejemplo que demuestra que c tiene una equivalencia de tipos por nombre,
   * no importa si dos tipos tienen la misma estructura, solo seran equivalentes
   * si tienen el mismo nombre. Es por ello que a una coordenada (c) no le puedo 
   * asignar un punto (p1), pese a que Punto y Coordenada tengan la misma estructura
   */
  return 0;
}
